### PR TITLE
change state prepared and delivered in one user action, closes #257

### LINKED ISSUE #257: Invert "mark as delivered"
At least for one organization, the DEFAULT is to always deliver the product before making the invoice.

This means that remembering to click "delivered" is an extra step that is easily forgotten.

The system should rather ask "at what date was this product delivered" when making the invoice, and default to "today".
There must also be a button that says something like "The product have not been delivered yet"

### PR DESCRIPTION
**Pull request by maxhansson**

### CODE REVIEW

File: app/decorators/sale_decorator.rb [hunk 1]
```diff
@@ -112,4 +112,7 @@ def user_email
     object.user.email
   end
 
+  def report_delivery
+    "report_delivery"
```
Comment (qzio) [inline app/decorators/sale_decorator.rb]: How is this used?
I suspect in a view somewhere? Will it be translated?

File: app/models/sale.rb [hunk 2]
```diff
@@ -168,6 +169,12 @@ def accounts_receivable_reverse_event
     end
   end
 
+  def report_delivery(transition)
+    if transition.args[1] == '1'
```
Comment (qzio) [inline app/models/sale.rb]: What does this mean?
Maybe a comment can be added to the code, describing why [1] have to be '1' etc.

File: app/models/sale.rb [hunk 3]
```diff
@@ -168,6 +169,12 @@ def accounts_receivable_reverse_event
...
+      send('deliver', transition.args[0])
```
Comment (qzio) [inline app/models/sale.rb]: it checks if .args[1] is 1 and then sends .args[0] to the deliver method?

File: app/views/sales/_action_buttons.html.haml [hunk 4]
```diff
@@ -9,7 +9,7 @@
       - if @sale.can_edit_items?
         %button{:class=>'btn btn-primary', :ng_click=>"open($event, 'sm','salesContent' )"}
           = t(:ordered)
-        = render partial: 'shared/modal_state_change', locals: {idx: 'salesContent', model: @sale, form_path: state_change_sale_path(@sale), event: 'mark_prepared', return_path: sale_path(@sale)}
+        = render partial: 'shared/modal_state_change', locals: {idx: 'salesContent', model: @sale, form_path: state_change_sale_path(@sale), event: 'mark_prepared', return_path: sale_path(@sale), p2: :report_delivery}
```
Comment (qzio) [inline app/views/sales/_action_buttons.html.haml]: what is `p2` ?

File: app/views/shared/_modal_state_change.html.haml [hunk 5]
```diff
@@ -1,4 +1,5 @@
 - p1 ||= nil
+- p2 ||= nil
```
Comment (qzio) [inline app/views/shared/_modal_state_change.html.haml]: What is p1 and p2 supposed to represent?
Can we come up with more descriptive names?

Comment (maxhansson) [inline app/views/shared/_modal_state_change.html.haml]: shared/model_state_change has two incoming parameters. 
p1 is an input-field and p2 is a checkbox.  A shared view with general fields. 
P1 may represent different input fields and P2 can be used for any checkbox.

File: app/views/sales/_action_buttons.html.haml (hunk 4 above)
Comment (maxhansson) [inline app/views/sales/_action_buttons.html.haml]: P2 is the variablename in the shared view. In this case it represent a checkbox for report_delivery.

File: app/models/sale.rb (hunk 2 above)
Comment (maxhansson) [inline app/models/sale.rb]: arg[1] = '1' is the checkbox for report delivery true.

File: app/models/sale.rb (hunk 3 above)
Comment (maxhansson) [inline app/models/sale.rb]: args[0] is the date of the state change.

File: app/decorators/sale_decorator.rb (hunk 1 above)
Comment (maxhansson) [inline app/decorators/sale_decorator.rb]: It is used in the view _action_buttons.html.haml
Translated in the view shared/model_state_change

File: app/models/sale.rb (hunk 2 above)
Comment (qzio) [inline app/models/sale.rb]: Can this be described in a comment.

For one that only looks at the code "transition.args[1]" doesn't mean anything.

File: app/models/sale.rb (hunk 3 above)
Comment (qzio) [inline app/models/sale.rb]: This either needs code comments, or better naming of the different args values.

File: app/views/sales/_action_buttons.html.haml (hunk 4 above)
Comment (qzio) [inline app/views/sales/_action_buttons.html.haml]: Would it be possible to use a name that describes what it is referring to?

p2 doesn't mean anything for someone just reading the source code.

File: app/models/sale.rb (hunk 2 above)
Comment (maxhansson) [inline app/models/sale.rb]: comments added 5b1f75a5cb8b3707025175667a81051a0c5b3cdc

File: app/models/sale.rb (hunk 3 above)
Comment (maxhansson) [inline app/models/sale.rb]: comments added 5b1f75a5cb8b3707025175667a81051a0c5b3cdc

File: app/views/sales/_action_buttons.html.haml (hunk 4 above)
Comment (maxhansson) [inline app/views/sales/_action_buttons.html.haml]: Would it be better with names from the tags type? p1 = input and p2 = checkbox?

Comment (qzio) [inline app/views/sales/_action_buttons.html.haml]: Yes, much better, makes it easier for someone like me only glances as the code.

**Pull request merged**